fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/cursors/SortCursorTests.java

```diff
@@ -0,0 +1,398 @@
+import java.io.File;
+import java.io.IOException;
+import java.security.SecureRandom;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Random;
+import java.util.function.Function;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+/**
+ * Tests for {@link MemorySortCursor} and {@link FileSortCursor}.
+ */
+@Tag(Tags.RequiresFDB)
+public class SortCursorTests extends FDBRecordStoreTestBase {
+    final KeyExpression num2Field = Key.Expressions.field("num_value_2");
+
+    private SortedRecordSerializer<Message> serializer;
+
+    private List<Integer> sortedNums;
+
+    @BeforeEach
+    public void setupRecords() throws Exception {
+        sortedNums = new ArrayList<>(100);
+        Random random = new Random(3456);
+        try (FDBRecordContext context = openContext()) {
+            openSimpleRecordStore(context);
+
+            serializer = new SortedRecordSerializer<>(DynamicMessageRecordSerializer.instance(), recordStore.getRecordMetaData(), recordStore.getTimer());
+
+            for (int i = 0; i < 100; i++) {
+                int num = random.nextInt();
+                sortedNums.add(num);
+
+                TestRecords1Proto.MySimpleRecord.Builder record = TestRecords1Proto.MySimpleRecord.newBuilder();
+                record.setRecNo(i);
+                record.setNumValue2(num);
+                recordStore.saveRecord(record.build());
+            }
+            commit(context);
+        }
+        Collections.sort(sortedNums);
+    }
+
+    abstract class MemoryAdapterBase implements MemorySortAdapter<Tuple, FDBStoredRecord<Message>> {
+        @Override
+        public int compare(Tuple o1, Tuple o2) {
+            return o1.compareTo(o2);
+        }
+
+        @Nonnull
+        @Override
+        public Tuple generateKey(FDBStoredRecord<Message> record) {
+            return num2Field.evaluateSingleton(record).toTuple();
+        }
+
+        @Nonnull
+        @Override
+        public byte[] serializeKey(final Tuple key) {
+            return key.pack();
+        }
+
+        @Override
+        public boolean isSerializedOrderReversed() {
+            return false;
+        }
+
+        @Nonnull
+        @Override
+        public Tuple deserializeKey(@Nonnull final byte[] key) {
+            return Tuple.fromBytes(key);
+        }
+
+        @Nonnull
+        @Override
+        public byte[] serializeValue(final FDBStoredRecord<Message> record) {
+            return serializer.serialize(record);
+        }
+
+        @Nonnull
+        @Override
+        public FDBStoredRecord<Message> deserializeValue(@Nonnull final byte[] bytes) {
+            return serializer.deserialize(bytes).getStoredRecord();
+        }
+
+        @Nonnull
+        @Override
+        public MemorySorter.SizeLimitMode getSizeLimitMode() {
+            return MemorySorter.SizeLimitMode.DISCARD;
+        }
+    }
+
+    @Test
+    public void memorySort() throws Exception {
```

Comment:
There are obviously a number of other tests that will ultimately need to be written (or, i suppose, be implicitly tested as part of actually executing and planning queries, but I think I'm more inclined to directly test behavior), such as reverse sorts, limit modes, and behavior in the face of various types of limits, but this is probably sufficient for prototyping.  I just don't want to lose track of the fact that work needs to be done here.

Reply:
I will second that, as a prototype, this is probably sufficient testing, but there are definitely a lot of additional tests of the cursors, and probably of the other components that should be added as this progresses.